rr_graph_uxsdcxx.cpp

```diff
@@ -0,0 +1,3174 @@
+			break;
+		default: break; /* Not possible. */
+		}
+	}
+	std::bitset<5> test_astate = astate | std::bitset<5>(0b00000);
+	if(!test_astate.all()) attr_error(test_astate, atok_lookup_t_channel);
+}
+
+void load_x_list(const pugi::xml_node &root, t_x_list *out){
+	if(root.first_child().type() == pugi::node_element)
+		throw std::runtime_error("Unexpected child element in <x_list>.");
+
+	std::bitset<2> astate = 0;
+	for(pugi::xml_attribute attr = root.first_attribute(); attr; attr = attr.next_attribute()){
+		atok_t_x_list in = alex_t_x_list(attr.name());
+		if(astate[(int)in] == 0) astate[(int)in] = 1;
+		else throw std::runtime_error("Duplicate attribute " + std::string(attr.name()) + " in <x_list>.");
+		switch(in){
+		case atok_t_x_list::INDEX:
+			out->index = std::strtoul(attr.value(), NULL, 10);
+			if(errno != 0)
+				throw std::runtime_error("Invalid value `" + std::string(attr.value()) + "` to load a unsigned int into out->index.");
+			break;
+		case atok_t_x_list::INFO:
+			out->info = std::strtol(attr.value(), NULL, 10);
+			if(errno != 0)
+				throw std::runtime_error("Invalid value `" + std::string(attr.value()) + "` to load a int into out->info.");
+			break;
+		default: break; /* Not possible. */
+		}
+	}
+	std::bitset<2> test_astate = astate | std::bitset<2>(0b00);
+	if(!test_astate.all()) attr_error(test_astate, atok_lookup_t_x_list);
+}
+
+void load_y_list(const pugi::xml_node &root, t_y_list *out){
+	if(root.first_child().type() == pugi::node_element)
+		throw std::runtime_error("Unexpected child element in <y_list>.");
+
+	std::bitset<2> astate = 0;
+	for(pugi::xml_attribute attr = root.first_attribute(); attr; attr = attr.next_attribute()){
+		atok_t_y_list in = alex_t_y_list(attr.name());
+		if(astate[(int)in] == 0) astate[(int)in] = 1;
+		else throw std::runtime_error("Duplicate attribute " + std::string(attr.name()) + " in <y_list>.");
+		switch(in){
+		case atok_t_y_list::INDEX:
+			out->index = std::strtoul(attr.value(), NULL, 10);
+			if(errno != 0)
+				throw std::runtime_error("Invalid value `" + std::string(attr.value()) + "` to load a unsigned int into out->index.");
+			break;
+		case atok_t_y_list::INFO:
+			out->info = std::strtol(attr.value(), NULL, 10);
+			if(errno != 0)
+				throw std::runtime_error("Invalid value `" + std::string(attr.value()) + "` to load a int into out->info.");
+			break;
+		default: break; /* Not possible. */
+		}
+	}
+	std::bitset<2> test_astate = astate | std::bitset<2>(0b00);
+	if(!test_astate.all()) attr_error(test_astate, atok_lookup_t_y_list);
+}
+
+void load_timing(const pugi::xml_node &root, t_timing *out){
+	if(root.first_child().type() == pugi::node_element)
+		throw std::runtime_error("Unexpected child element in <timing>.");
+
+	std::bitset<5> astate = 0;
+	for(pugi::xml_attribute attr = root.first_attribute(); attr; attr = attr.next_attribute()){
+		atok_t_timing in = alex_t_timing(attr.name());
+		if(astate[(int)in] == 0) astate[(i
```

Comment:
Why `[2][1]`? Add some comments or don't use magic numbers?

Reply:
Comments are a good idea. I should write something like an `annotate_state_machine` function.

Just for the record, `[2][1]` means there are 2 possible states and 1 possible input. Two rows of `{0}` means that both states go to state 0 when input is 0 (`gtok_t_metadata::META`). A few lines below, there is the `if(state != 0)` check. That means 0 is an accepted state. If the final state isn't 0, we call `dfa_error`, which throws with an error message `Expected meta, found end of input`.

How can I annotate this? There is no way with input token names, since C++ doesn't support array initializations with enum indices. ([C does!](https://eli.thegreenplace.net/2011/02/15/array-initialization-with-enum-indices-in-c-but-not-c)). I think I can name the states after their expected inputs, like:
```c++
enum class metadata_state = { INVALID = -1, EXPECT_META, EXPECT_META_OR_EOF };
metadata_state gstate_t_metadata[2][1] = {
    { EXPECT_META_OR_EOF },
    { EXPECT_META_OR_EOF },
};
...
    metadata_state next, state = metadata_state::EXPECT_META;
...
    next = gstate_t_metadata[(int)state][(int)in];
    if(next == INVALID)
...
    if(state != metadata_state::EXPECT_META_OR_EOF)
...
```

Reply:
As C++ doesn't support enum initializers, you could try something like;
```cpp
enum Fruit_t {
    APPLES = 1,
    ORANGES = 2,
    STRAWBERRIES = 3
};

void foo()
{
    static const int price_lookup[] = {
        [1 /* APPLES */] = 6,
        [2 /* ORANGES */] = 10,
        [3 /* STRAWBERRIES */] = 55
    };
}
```

Maybe the following;
```cpp
metadata_state gstate_t_metadata[2 /* Number states */][1 /* Number possible input */] = ....
```
or
```cpp
#define NUM_METADATA_STATES 2
#define NUM_METADATA_INPUTS 1

metadata_state gstate_t_metadata[NUM_METADATA_STATES][NUM_METADATA_INPUTS] = ....
```

Reply:
I implemented the former, but turns out C++ actually doesn't support it and causes `-Wpedantic` to fail: https://stackoverflow.com/questions/18731707/why-does-c11-not-support-designated-initializer-lists-as-c99

It looks like C++ really didn't try to have convenient N dimensional arrays.